src/passes/OptimizeInstructions.cpp

```diff
@@ -3597,6 +3597,40 @@ struct OptimizeInstructions
         }
       }
     }
+    {
+      //   x !=  NaN   ==>   1
+      //   x <=> NaN   ==>   0
+      //   x op  NaN'  ==>   NaN',  iff `op` != `copysign` and `x` != C
+      Const* c;
+      Binary* bin;
+      Expression* x;
+      if (matches(curr, binary(&bin, pure(&x), fval(&c))) &&
+          std::isnan(c->value.getFloat()) && !x->is<Const>() &&
```

Comment:
I usually skipped the constant to the right side for the floating point because there could be NaN, which was better to handle in propagate constant pass. I think since OptimizeInstruction now uses the same utility for nan-canonicalization we can remove this restriction.